core/src/main/java/org/infinispan/scattered/impl/ScatteredStateConsumerImpl.java

```diff
@@ -503,6 +503,11 @@ private void applyValues(Address address, List<Object> keys, Response response)
       for (int i = 0; i < keys.size(); ++i) {
          Object key = keys.get(i);
          InternalCacheValue icv = values[i];
+         if (icv == null) {
+            // The entry got lost in the meantime - this can happen when the cache is stopped concurrently to processing
+            // the GetAllCommand. We'll just avoid NPEs here: data is lost as > 1 nodes have left.
```

Comment:
Shouldn't we get an exception if the cache is being stopped?

Reply:
Not when the command enters interceptor stack and the cache is stopped only after that.

Reply:
If it can happen for our internal `GetAllCommand` it can probably happen for the application's `GetAllCommands` as well, so I'd change the code to throw an `IllegalLifecycleStateException` if possible.

Reply:
On the stopping receiver this topology check was ignored because the GetAllCommand has `SKIP_OWNERSHIP_CHECK` during the state transfer - this node was *not* the owner in previous topology (so it can't respond according to CH) but it has responded during the keys transfer as having the last version of the key. However as it's being stopped the data-container has been cleared concurrently and it fetches null data.
I agree that this could be improved by another type of signalization but this is out of scope of this PR.
On the originator, I don't see why you should throw an exception instead of processing the other data. Would you be ok with issuing a WARN log here?

Reply:
You're probably right that this is out of the scope of this PR, we could handle ISPN-9095 separately.

> However as it's being stopped the data-container has been cleared concurrently and it fetches null data.

When a cache shuts down while a remote command is running I would expect `BaseBlockingRunnable` to replace the command's response with a `CacheNotFoundResponse`. `DataContainer.clear()` has priority 999 and `PerCacheInboundInvocationHandler.stop()` has priority 10, so IMO the `GetAllCommand` should return either the proper value or a CNFR. BTW, non-scattered caches also rely on CNFR to know that a remote node is no longer an owner and the originator should ask another node, so if the CNFR logic is broken it's even more important to fix it.

> On the originator, I don't see why you should throw an exception instead of processing the other data. Would you be ok with issuing a WARN log here?

I'd fix the remote node to properly reply with a CNFR when it's shutting down, and log an error in `ScatteredStateConsumerImpl` if a value was invalid. If SSCI can retry that key and fetch the value from another node we could make it a warning instead.

Reply:
Oh, I was wrong... I thought that the test is already stopping the cache after test failure but in fact the `TestingUtil` is clearing the data container directly. So this was a test problem.

In this case I don't understand why the exception happened in https://issues.jboss.org/browse/ISPN-9095

Reply:
Ok, then do we still need this change?